From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Wed, 29 Oct 2025 11:45:23 +0000
Subject: chore: handle support for `import defer * as ns` and `import.defer`
 syntax

V8 added support for the above syntax https://chromium-review.googlesource.com/c/v8/v8/+/7017517
by adding a new `ModuleImportPhase::kDefer` phase.

This patch can be removed when Electron updates to a version of Node.js containing
the above CL.

diff --git a/src/module_wrap.cc b/src/module_wrap.cc
index a584e3a80adb69d2028dc79450349823ab973a58..6f010fa2e014b2d13b1f89a691d09e2ffdf690b6 100644
--- a/src/module_wrap.cc
+++ b/src/module_wrap.cc
@@ -563,8 +563,10 @@ ModulePhase to_phase_constant(ModuleImportPhase phase) {
       return kEvaluationPhase;
     case ModuleImportPhase::kSource:
       return kSourcePhase;
+    case ModuleImportPhase::kDefer:
+    default:
+      UNREACHABLE();
   }
-  UNREACHABLE();
 }
 
 static Local<Object> createImportAttributesContainer(
@@ -1471,6 +1473,7 @@ void ModuleWrap::CreatePerContextProperties(Local<Object> target,
 
   V(ModulePhase, kEvaluationPhase);
   V(ModulePhase, kSourcePhase);
+  V(ModulePhase, kDeferPhase);
 #undef V
 }
 
diff --git a/src/module_wrap.h b/src/module_wrap.h
index 45264c2ad58e37a73fd62addac7fb671eed6d502..fb32fbc5c1cfa4aef4a7822dbc6195429299da3e 100644
--- a/src/module_wrap.h
+++ b/src/module_wrap.h
@@ -37,6 +37,7 @@ enum HostDefinedOptions : int {
 enum ModulePhase : int {
   kSourcePhase = 1,
   kEvaluationPhase = 2,
+  kDeferPhase = 3,
 };
 
 /**
